File: 2015/21/boss.cpp
#include "boss.h"

Boss::Boss(int points, int damage, int armor)
  : _points(points)
  , _damage(damage)
  , _armor(armor)
{}

bool attack(int& points, int damage, int armor)
{
  if(armor > damage)
    --points;
  else
    points -= (damage - armor);

  return points <= 0;
}

int Boss::play(int points, int damage, int armor) const
{
  int bosspoints = _points;
  for(int i = 0; ; ++i)
  {
    if(i % 2 == 0)
    {
      if(attack(bosspoints, damage, _armor))
        break;
    }
    else
    {
      if(attack(points, _damage, armor))
        break;
    }
  }

  return points - bosspoints;
}

File: 2015/22/boss.rb
require_relative 'player'

class Boss < Player
  attr_reader :damage
  def initialize(points, damage)
    super points
    @damage = damage
  end

  def to_s
    "Boss points=#{points}"
  end

  def win
    nil
  end

  def before_turn!
    true
  end

  def take_turn(spent, effects, time, &block)
    attack = Attack.new time, damage
    block.call attack
  end

  class Attack
    attr_reader :name, :damage
    def initialize(time, damage)
      @time = time
      @name = 'boss attack'
      @damage = damage
    end
    #def active_at?(time)
      #@time == time
    #end
  end
end
